include/aspect/material_model/diffusion_dislocation.h

```diff
@@ -0,0 +1,200 @@
+         */
+        virtual bool
+        compressibility_depends_on (const NonlinearDependence::Dependence dependence) const;
+
+        /**
+         * Return true if the specific_heat() function returns something that
+         * may depend on the variable identifies by the argument.
+         */
+        virtual bool
+        specific_heat_depends_on (const NonlinearDependence::Dependence dependence) const;
+
+        /**
+         * Return true if the thermal_conductivity() function returns
+         * something that may depend on the variable identifies by the
+         * argument.
+         */
+        virtual bool
+        thermal_conductivity_depends_on (const NonlinearDependence::Dependence dependence) const;
+
+        /**
+         * Return whether the model is compressible or not.  Incompressibility
+         * does not necessarily imply that the density is constant; rather, it
+         * may still depend on temperature or pressure. In the current
+         * context, compressibility means whether we should solve the contuity
+         * equation as $\nabla \cdot (\rho \mathbf u)=0$ (compressible Stokes)
+         * or as $\nabla \cdot \mathbf{u}=0$ (incompressible Stokes).
+         */
+        virtual bool is_compressible () const;
+
+        virtual double reference_viscosity () const;
+
+        virtual double reference_density () const;
+
+        static
+        void
+        declare_parameters (ParameterHandler &prm);
+
+        virtual
+        void
+        parse_parameters (ParameterHandler &prm);
+
+      private:
+
+        double reference_T;
+        double min_strain_rate;
+        double min_visc;
+        double max_visc;
+        double veff_coefficient;
+        double ref_visc;
+        double thermal_diffusivity;
+        double heat_capacity;
+
+        /**
+         * From multicomponent material model: From a list of compositional
+         * fields of length N, we come up with an N+1 length list that which
+         * also includes the fraction of ``background mantle''. This list
+         * should sum to one, and is interpreted as volume fractions.  If the
+         * sum of the compositional_fields is greater than one, we assume that
+         * there is no background mantle (i.e., that field value is zero).
+         * Otherwise, the difference between the sum of the compositional
+         * fields and 1.0 is assumed to be the amount of background mantle.
+         */
+        std::vector<double> compute_volume_fractions(
+          const std::vector<double> &compositional_fields) const;
+        std::vector<double> densities;
+        std::vector<double> thermal_expansivities;
+
+        /**
+         * Enumeration for selecting which viscosity averaging scheme to use.
+         * Select between harmonic, arithmetic, geometric, and
+         * maximum_composition.  The max composition scheme simply uses the
+         * viscosity of whichever field has the highes volume fraction.
```

Comment:
what does it average over? the compositionally dependent values, or the quadrature points?